src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java

```diff
@@ -0,0 +1,589 @@
+                                                            .iterator(),
+                                                   decoratedKey ->
+                                                   {
+                                                       K partitionKey = compose(converters.get(partitionKeyClass.get()),
+                                                                                decoratedKey.getKey());
+                                                       boolean keyRequired = mapKeyFilter.test(partitionKey);
+                                                       if (!keyRequired)
+                                                           return null;
+
+                                                       C value = map.get(partitionKey);
+                                                       return mapValueFilter.test(value) ? rowConverter.apply(
+                                                           partitionKey, value) : null;
+                                                   });
+    }
+
+    public static String virtualTableNameStyle(String camel)
+    {
+        // Process sub names in the full metrics group name separately and then join them.
+        // For example: "ClientRequest.Write-EACH_QUORUM" will be converted to "client_request_write_each_quorum".
+        String[] subNames = ONLY_ALPHABET_PATTERN.matcher(camel).replaceAll(".").split("\\.");
+        return Arrays.stream(subNames)
+                     .map(CollectionVirtualTableAdapter::camelToSnakeWithAbbreviations)
+                     .reduce((a, b) -> a + '_' + b)
+                     .orElseThrow(() -> new IllegalArgumentException("Invalid table name: " + camel));
+    }
+
+    private static String camelToSnakeWithAbbreviations(String camel)
+    {
+        Pattern pattern = Pattern.compile("^[A-Z1-9_]+$");
+        // Contains only uppercase letters, numbers and underscores, so it's already snake case.
+        if (pattern.matcher(camel).matches())
+            return camel.toLowerCase();
+
+        // Some special cases must be handled manually.
+        String modifiedCamel = camel;
+        for (Pair<String, String> replacement : knownAbbreviations)
+            modifiedCamel = modifiedCamel.replace(replacement.left, replacement.right);
+
+        return camelToSnake(modifiedCamel);
+    }
+
+    private TableMetadata buildMetadata(String keyspaceName, String tableName, String description, RowWalker<R> walker)
+    {
+        TableMetadata.Builder builder = TableMetadata.builder(keyspaceName, tableName)
+                                                     .comment(description)
+                                                     .kind(TableMetadata.Kind.VIRTUAL);
+
+        List<AbstractType<?>> partitionKeyTypes = new ArrayList<>(walker.count(Column.Type.PARTITION_KEY));
+        walker.visitMeta(
+            new RowWalker.MetadataVisitor()
```

Comment:
should not this be again aligned? it should start on the same line as visitMeta

Reply:
Fixed.